src/storage/readpool_context.rs

```diff
@@ -0,0 +1,146 @@
+        Context {
+            pd_sender,
+            command_duration: SCHED_HISTOGRAM_VEC.local(),
+            processing_read_duration: SCHED_PROCESSING_READ_HISTOGRAM_VEC.local(),
+            command_keyreads: KV_COMMAND_KEYREAD_HISTOGRAM_VEC.local(),
+            kv_command_counter: KV_COMMAND_COUNTER_VEC.local(),
+            raw_command_counter: RAWKV_COMMAND_COUNTER_VEC.local(),
+            command_pri_counter: SCHED_COMMANDS_PRI_COUNTER_VEC.local(),
+            scan_details: KV_COMMAND_SCAN_DETAILS.local(),
+            read_flow_stats: HashMap::default(),
+        }
+    }
+
+    #[inline]
+    pub fn collect_command_count_and_duration(
+        &mut self,
+        cmd: &str,
+        priority: readpool::Priority,
+        is_raw: bool,
+    ) -> LocalHistogramTimer {
+        if is_raw {
+            self.raw_command_counter.with_label_values(&[cmd]).inc();
+        } else {
+            self.kv_command_counter.with_label_values(&[cmd]).inc();
+        }
+        self.command_pri_counter
+            .with_label_values(&[&priority.to_string()])
+            .inc();
+        self.command_duration
+            .with_label_values(&[cmd])
+            .start_coarse_timer()
+    }
+
+    #[inline]
+    pub fn collect_processing_read_duration(&mut self, cmd: &str) -> LocalHistogramTimer {
+        self.processing_read_duration
+            .with_label_values(&[cmd])
+            .start_coarse_timer()
+    }
+
+    #[inline]
+    pub fn collect_key_reads(&mut self, cmd: &str, count: u64) {
+        self.command_keyreads
+            .with_label_values(&[cmd])
+            .observe(count as f64);
+    }
+
+    #[inline]
+    pub fn collect_scan_count(&mut self, cmd: &str, statistics: &storage::Statistics) {
+        for (cf, details) in statistics.details() {
+            for (tag, count) in details {
+                self.scan_details
+                    .with_label_values(&[cmd, cf, tag])
+                    .inc_by(count as f64)
+                    .unwrap();
+            }
+        }
+    }
+
+    #[inline]
+    pub fn collect_read_flow(&mut self, region_id: u64, statistics: &storage::Statistics) {
+        if self.pd_sender.is_none() {
+            return;
+        }
+        let flow_stats = self.read_flow_stats
+            .entry(region_id)
+            .or_insert_with(storage::FlowStatistics::default);
+        flow_stats.add(&statistics.write.flow_stats);
+        flow_stats.add(&statistics.data.flow_stats);
+    }
+}
+
+impl futurepool::Context for Context {
+    fn on_tick(&mut self) {
+        // Flush Prometheus metrics
+        self.command_duration.flush();
+        self.processing_read_duration.flush();
+        self.command_keyreads.flush();
+        self.kv_command_counter.flush();
+        self.raw_command_counter.flush();
+        self.command_pri_counter.flush();
+        self.scan_details.flush();
+
+        // Report PD metrics
+        if !self.read_flow_stats.is_empty() {
```

Comment:
if empty or pd_sender is none, return directly?

Reply:
we may report more things to pd in the future, so that I prefer `if ... {}` to `if ! ...  { return; }` here :)